From 8100a635b6b2661a7a4fd9c5ca86876805992413 Mon Sep 17 00:00:00 2001
From: Martin Blumenstingl <martin.blumenstingl@googlemail.com>
Date: Tue, 2 Jul 2019 00:07:57 +0200
Subject: [PATCH 1/5] dt-bindings: pci: document the Lantiq VRX200/ARX300 PCIe
 controller

The Lantiq VRX200 SoCs embeds one PCIe controller which is based on the
Synopsys DesignWare PCIe core. The ARX300 SoC embeds two of the same
PCIe controllers.

The controller itself is part of the "fpi" bus using the "dbi" register
space (which also contains the DesignWare PCIe controller registers) and
an "app" register space which is a custom extension. There is a PCIe PHY
which is not part of the PCIe controller register space - instead it is
located on the RCU bus.

Like various other IP blocks on these SoCs the endianness is configured
in a special RCU register.

Signed-off-by: Martin Blumenstingl <martin.blumenstingl@googlemail.com>
---
 .../devicetree/bindings/pci/lantiq,pcie.yaml  | 190 ++++++++++++++++++
 1 file changed, 190 insertions(+)
 create mode 100644 Documentation/devicetree/bindings/pci/lantiq,pcie.yaml

--- /dev/null
+++ b/Documentation/devicetree/bindings/pci/lantiq,pcie.yaml
@@ -0,0 +1,190 @@
+# SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/phy/lantiq,vrx200-pcie-phy.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: Lantiq PCIe Controller Device Tree Bindings
+
+maintainers:
+  - Martin Blumenstingl <martin.blumenstingl@googlemail.com>
+
+properties:
+
+  compatible:
+    oneOf:
+      - items:
+          - enum:
+              - lantiq,vrx200-pcie
+              - lantiq,arx300-pcie
+          - const: lantiq,dw-pcie
+          - const: snps,dw-pcie
+
+  reg:
+    items:
+      - description: The DesignWare PCIe controller's DBI register space
+      - description: The PCI configuration space
+      - description: The Lantiq PCIe Controller application register space
+
+  reg-names:
+    items:
+      - const: dbi
+      - const: config
+      - const: app
+
+  ranges: true
+
+  "#address-cells":
+    enum: [ 3 ]
+
+  "#size-cells":
+    enum: [ 2 ]
+
+  device_type:
+    items:
+      - const: pci
+
+  interrupts:
+    items:
+      - description: The PCIe Controller IR interrupt line
+
+  interrupt-names:
+    items:
+      - const: ir
+
+  "#interrupt-cells":
+    description: TODO
+
+  interrupt-map:
+    description: TODO
+
+  interrupt-map-mask:
+    description: TODO
+
+  lantiq,rcu:
+    $ref: /schemas/types.yaml#/definitions/phandle
+    description: phandle to the RCU syscon
+
+  lantiq,rcu-endian-offset:
+    $ref: /schemas/types.yaml#/definitions/uint32
+    description: the offset of the endian registers for this PHY instance in the RCU syscon
+
+  lantiq,rcu-inbound-big-endian-mask:
+    $ref: /schemas/types.yaml#/definitions/uint32
+    description: the mask to control the inbound endianness conversion
+
+  lantiq,rcu-outbound-big-endian-mask:
+    $ref: /schemas/types.yaml#/definitions/uint32
+    description: the mask to control the output endianness conversion
+
+  big-endian:
+    description: Configures the PDI (PHY) registers in big-endian mode
+    type: boolean
+
+  little-endian:
+    description: Configures the PDI (PHY) registers in big-endian mode
+    type: boolean
+
+  resets:
+    maxItems: 1
+    description: phandle to the reset line which is shared between the PCIe core and PHY
+
+  clocks:
+    items:
+      - description: the PCIe controller module clock
+      - description: the PCIe bus clock
+      - description: the AHB clock
+
+  clock-names:
+    items:
+      - const: ctrl
+      - const: pcie_bus
+      - const: ahb
+
+  phys:
+    $ref: /schemas/types.yaml#/definitions/phandle
+    description: phandle to the PCIe PHY
+
+  phy-names:
+    items:
+      - const: pcie
+
+  num-lanes:
+    $ref: /schemas/types.yaml#/definitions/uint32
+    description: the number of PCIe lanes which are used by this controller
+
+  reset-gpios:
+    maxItems: 1
+    description: phandle to the reset GPIO, see designware-pcie.txt
+
+
+required:
+  - compatible
+  - reg
+  - reg-names
+  - ranges
+  - "#address-cells"
+  - "#size-cells"
+  - device_type
+  - interrupts
+  - interrupt-names
+  - "#interrupt-cells"
+  - lantiq,rcu
+  - lantiq,rcu-endian-offset
+  - lantiq,rcu-inbound-big-endian-mask
+  - lantiq,rcu-outbound-big-endian-mask
+  - resets
+  - clocks
+  - clock-names
+  - phys
+  - phy-names
+
+
+additionalProperties: true
+
+examples:
+  - |
+    #include <dt-bindings/gpio/gpio.h>
+
+    pcie: pcie@d900000 {
+        compatible = "lantiq,xrx200-pcie", "snps,dw-pcie";
+
+        reg = <0xd900000 0x1000>,
+              <0xd000000 0x800000>,
+              <0xe100900 0x200>;
+        reg-names = "dbi", "config", "app";
+
+        ranges = <0x02000000 0 0         0xc000000 0 0x1000000
+                  0x01000000 0 0xd800000 0xd800000 0 0x0100000>;
+
+        #address-cells = <3>;
+        #size-cells = <2>;
+        device_type = "pci";
+
+        interrupts = <161>;
+        interrupt-names = "ir";
+        #interrupt-cells = <1>;
+        interrupt-map-mask = <0 0 0 0x7>;
+        interrupt-map = <0 0 0 1 &icu0 144>, /* int_a */
+                        <0 0 0 2 &icu0 145>, /* int_b */
+                        <0 0 0 3 &icu0 146>, /* int_c */
+                        <0 0 0 4 &icu0 147>; /* int_d */
+
+        lantiq,rcu = <&rcu0>;
+        lantiq,rcu-endian-offset = <0x4c>;
+        lantiq,rcu-inbound-big-endian-mask = <0x2>; /* bit 1 */
+        lantiq,rcu-outbound-big-endian-mask = <0x14>; /* bit 4 */
+        big-endian;
+
+        resets = <&reset0 22 22>;
+
+        clocks = <&pmu 33>, <&pmu 31>, <&pmu 13>;
+        clock-names = "ctrl", "pcie_bus", "ahb";
+
+        phys = <&pcie_phy>;
+        phy-names = "pcie";
+
+        num-lanes = <1>;
+
+        reset-gpios = <&gpio 38 GPIO_ACTIVE_LOW>;
+    };
